fe/fe-core/src/main/java/org/apache/doris/rewrite/InferFiltersRule.java

```diff
@@ -0,0 +1,302 @@
+            SlotRef leftSlot = conjunct.getChild(0).unwrapSlotRef();
+            SlotRef rightSlot = conjunct.getChild(1).unwrapSlotRef();
+
+            if (checkSlot.simple_equals(leftSlot)) {
+                addNewBinaryPredicate(genNewBinaryPredicate(analyzer, slotToLiteral, rightSlot),
+                        slotToLiteralPredicatesSet, newPredicatesList,
+                        tmpSlotToLiteralPredicatesList,
+                        isNeedInfer(checkSlot, rightSlot));
+            } else if (checkSlot.simple_equals(rightSlot)) {
+                addNewBinaryPredicate(genNewBinaryPredicate(analyzer, slotToLiteral, leftSlot),
+                        slotToLiteralPredicatesSet, newPredicatesList,
+                        tmpSlotToLiteralPredicatesList,
+                        isNeedInfer(checkSlot, leftSlot));
+            }
+        }
+    }
+
+    private boolean isNeedInfer(SlotRef checkSlot, SlotRef newSlot) {
+        JoinOperator jop = newSlot.getDesc().getParent().getRef().getJoinOp();
+        boolean ret = false;
+        if (jop == null) {
+            JoinOperator checkJop = checkSlot.getDesc().getParent().getRef().getJoinOp();
+            if (checkJop.isInnerJoin()
+                    || checkJop == JoinOperator.RIGHT_SEMI_JOIN) {
+                ret = true;
+            } else if (checkJop == JoinOperator.RIGHT_OUTER_JOIN
+                    || checkJop == JoinOperator.RIGHT_ANTI_JOIN) {
+                ret = true;
+            }
+        } else if (jop.isInnerJoin()
+                || jop == JoinOperator.LEFT_SEMI_JOIN) {
+            ret = true;
+        } else if (jop == JoinOperator.LEFT_OUTER_JOIN
+                || jop == JoinOperator.LEFT_ANTI_JOIN) {
+            ret = true;
+        }
+        return ret;
+    }
+
+    private Expr genNewBinaryPredicate(Analyzer analyzer, Expr oldExpr, Expr newSlot) {
+        if (oldExpr instanceof BinaryPredicate) {
+            BinaryPredicate oldBP = (BinaryPredicate) oldExpr;
+            BinaryPredicate newBP = new BinaryPredicate(oldBP.getOp(), newSlot, oldBP.getChild(1));
+            newBP.analyzeNoThrow(analyzer);
+            return newBP;
+        }
+        return oldExpr;
+    }
+
+    private void addNewBinaryPredicate(Expr expr,
+                                       Set<Pair<Expr, Expr>> slotToLiteralPredicatesSet,
+                                       ArrayList<Expr> newPredicatesList,
+                                       ArrayList<Expr> tmpSlotToLiteralPredicatesList,
+                                       boolean needAddNewPredicatesList) {
+        if (expr instanceof BinaryPredicate) {
+            BinaryPredicate newBP = (BinaryPredicate) expr;
+            Pair<Expr, Expr> pair = new Pair<>(newBP.getChild(0), newBP.getChild(1));
+            if (!slotToLiteralPredicatesSet.contains(pair)) {
+                slotToLiteralPredicatesSet.add(pair);
+                if (needAddNewPredicatesList) {
+                    newPredicatesList.add(newBP);
```

Comment:
Should the new predicate also be de-duplicated?
for example
a=b and b=c and a=1 and c=1
The b=1 may be inferred twice.